src/deepsparse/pipeline.py

```diff
@@ -176,22 +174,113 @@ def __init__(
+                f"{self.input_schema} object. Inputs parsed to {type(pipeline_inputs)}"
+            )
+
+        # batch size of the inputs may be `> self._batch_size` at this point
+        engine_inputs: List[numpy.ndarray] = self.process_inputs(pipeline_inputs)
+        if isinstance(engine_inputs, tuple):
+            engine_inputs, postprocess_kwargs = engine_inputs
         else:
-            # use passed in executor
-            executor = kwargs.pop("executor")
+            postprocess_kwargs = {}
+
+        # split inputs into batches of size `self._batch_size`
+        batches = self.split_engine_inputs(engine_inputs, self._batch_size)
+
+        # submit to engine
+        futures = [
+            self.executor.submit(self.engine_forward, batch) for batch in batches
+        ]
+        wait(futures)
+
+        # join together the batches of size `self._batch_size`
+        engine_outputs = self.join_engine_outputs(
+            [future.result() for future in futures]
+        )
 
-        return (
-            executor.submit(self._run, *args, **kwargs)  # Non-Blocking call
-            if executor and not self.use_dynamic_batch()
-            else self._run(*args, **kwargs)  # Blocking call
+        pipeline_outputs = self.process_engine_outputs(
+            engine_outputs, **postprocess_kwargs
         )
+        if not isinstance(pipeline_outputs, self.output_schema):
+            raise ValueError(
+                f"Outputs of {self.__class__} must be instances of "
+                f"{self.output_schema} found output of type {type(pipeline_outputs)}"
+            )
+
+        return pipeline_outputs
+
+    def split_engine_inputs(
+        self, items: List[numpy.ndarray], batch_size: int
+    ) -> List[List[numpy.ndarray]]:
+        """
+        Splits each item into numpy arrays with the first dimension == `batch_size`.
+
+        For example, if `items` has three numpy arrays with the following
+        shapes: `[(4, 32, 32), (4, 64, 64), (4, 128, 128)]`
+
+        Then with `batch_size==4` the output would be:
+        ```
+        [[(4, 32, 32), (4, 64, 64), (4, 128, 128)]]
+        ```
+
+        Then with `batch_size==2` the output would be:
+        ```
+        [
+            [(2, 32, 32), (2, 64, 64), (2, 128, 128)],
+            [(2, 32, 32), (2, 64, 64), (2, 128, 128)],
+        ]
+        ```
+
+        Then with `batch_size==1` the output would be:
+        ```
+        [
+            [(1, 32, 32), (1, 64, 64), (1, 128, 128)],
+            [(1, 32, 32), (1, 64, 64), (1, 128, 128)],
+            [(1, 32, 32), (1, 64, 64), (1, 128, 128)],
+            [(1, 32, 32), (1, 64, 64), (1, 128, 128)],
+        ]
+        ```
+        """
+        assert all(isinstance(item, numpy.ndarray) for item in items)
+
+        # all items should have the same batch size
+        total_batch_size = items[0].shape[0]
+        assert all(item.shape[0] == total_batch_size for item in items)
```

Comment:
nit: prefer if/raise w/ description (to avoid string building in optional assert error message)

Reply:
I added more detail comments around these asserts - they are things that indicate an internal error that should never happen, so there's not a helpful message we could send to the user